### DISCUSSION TITLE
Customizations with the @annotorious/react

### DISCUSSION THREAD

**Discussion by ghost** [annotorious/annotorious#331] [Q&A]
Firstly, it's nice to have a package that handles both the annotations and the viewer in react. I had a few things I wanted to ask and hopefully get some directions on how to integrate them. 

1. The library has a Rectangle as its default shape. I see there's "annotorious-v2-selector-pack" library which has got other shapes too. How can I select the shape (tool) with this package using the library and perhaps even create our own shapes? 

2. Is there also a way to "erase" the created shapes? 

3. It would be great to have an even greater control over the shapes (like the css).

The docs are probably a better place for some of these. In any case, appreciate the time and effort put into this library.

**Comment by rsimon** [comment]
Some answers below: 

- v2.x extensions will __no longer be compatible with v3__. Out of the box, Annotorious 3 has rectangle and polygon tools, like the old version. In addition, there's [a new selector pack in the works](https://github.com/annotorious/annotorious-v3-extensions/tree/main/annotorious-selector-pack) which currently adds a circle/ellipse tool.
- Internally, Annotorious is written in Svelte. You could gather some inspiration [here](https://github.com/annotorious/annotorious-v3-extensions/blob/main/annotorious-selector-pack/src/ellipse/RubberbandEllipse.svelte) and [here](https://github.com/annotorious/annotorious-v3-extensions/blob/main/annotorious-selector-pack/src/ellipse/EllipseEditor.svelte) on how to write new drawing tools and shape editors, respectively.
- In React you can generally change the tool with the `<ImageAnnotator tool={tool} />` property. But  documentation is still... ahm... waiting to be written ;-) Meanwhile, see [here for a real-world code example](https://github.com/rsimon/immarkus/blob/main/src/pages/annotate/WorkspaceSection/AnnotatableImage.tsx#L58-L81).  (This also uses the selector pack plugin, BTW.)
- There is currently no "controlled mode" for Annotorious React, or callback properties. This might come in the future. At the moment, you can get a reference to the `anno` object using the [useAnnotator hook](https://github.com/annotorious/annotorious/blob/3.x/packages/annotorious-react/src/Annotorious.tsx#L111-L114).
- The `anno` object offers largely the same API methods as v2. You can use `anno.removeAnnotation(id)` to remove an annotation.
- You can control the annotation style via the `<ImageAnnotator style={...} />` property. The style property takes either an object of type [DrawingStyle](https://github.com/annotorious/annotorious/blob/3.x/packages/annotorious-core/src/model/DrawingStyle.ts), or a __function__ that takes an annotation as argument and returns a DrawingStyle.

**Reply by ghost** [reply]
These are some quality resources, thanks! I was able to inject the plugin and get it working.

I also did notice some new and exciting stuff in the extensions repo. I see packages such as annotorious-comment-popup, annotorious-osd-textlayer and wanted to ask if there's a way to get them plugged in just like the Selectorpack. 

There was also a "freehand" tool in the v2 extensions. Is the approach to get this tool added going to be the same as with the custom shapes? I'm trying to draw parallels between the old code and the new one, but had no luck so far :')

**Reply by rsimon** [reply]
Yes, almost too much work in progress ;-) The annotorious-comment-popup isn't done yet. (Actually not even properly started...) The text layer plugin might work... Although I haven't tried it with React yet, only in a vanilla JS setup. (It should work in principle. But I wouldn't be surprised if it still lacks some plumbing to get it working.)

Some background:

- The `annotorious-comment-popup` is supposed to offer a backwards-compatible UI for those who are now using Annotorious v2 with the default comment/tag popup. I was never particularly fond of the default popup, and many people ended up not using it, anyway. Also: extending the default popup was a mess, especially if you wanted to use React. That's why Annotorious v3 will no longer have a popup at all. Instead, there will be helpers that make it easy to build your own popup in the framework of your choice.  __However:__ for those who just need a basic, non-extensible drop-in replacement for the old UI, the `annotorious-comment-popup` will be it.
- The `osd-text-layer` extension will be great :-) In spirit, it replaces the old [Shape Labels plugin](https://github.com/annotorious/annotorious-v2-plugins/tree/main/plugins/annotorious-shape-labels). But it's a lot more flexible, customizable, and will have better performance.
- A freehand tool is something I would like to build for v3, too. But it's not on the immediate roadmap. Which likely means I won't personally work on it any time soon - unless I happen to find a paying customer for it. (My goal would be to build a slightly enhanced version, that does polygon simplification as you draw.)
- I'm also still planning [to add a Photoshop-style](https://twitter.com/aboutgeo/status/1492485333082071044) smart [scissors tool](https://twitter.com/aboutgeo/status/1481302036721324038). But, likewise, probably something that has to wait until I actually find a paying customer (or end up having A LOT of spare time at some point ;-)
- Re-use of any of the old drawing tools, I'm afraid, is out of the question. v3 is a complete re-write that changed literally __everything__. So: no surprise that you had no luck drawing parallels between the old and the new code ;-)

**Reply by ghost** [reply]
This is great news, thanks for all the updates. Just wanted to ask/mention a couple of things.

I tried the command `npm install --save @annotorious/openseadragon-textlayer`, but it throws a "Not Found" error. 
 
I was also leaning towards making some changes/add new drawing tools as you mentioned by forking the repo. However, it looks like a simple npm install wasn't cutting it. It might have been due to some paths in the exports not being available. Do you have any advice on how I could resolve this?

**Reply by ghost** [reply]
I got the repo part (selector-pack) working.

**Comment by rsimon** [comment]
I think some of the work-in-progress repos might be pointing to super-early versions of Annotorious which I've unpublished in the meantime, to avoid having a list of 100 dev versions on npm. (You can see it if the version has something like `-pre-alpha` in it.) The versions that are published now are getting fairly stable. They all have `-rc.X` suffixes, and won't get unpublished. So, that's something you might want to check. 

As for the text layer specifically, I think I haven't actually published that so far. I can see that the dependencies still point to local folders. I.e. for now, you could check out the Annotorious 3 branch locally, and change the paths to match your own system. 

I'll try to roll an official release. But probably not this week anymore.

**Reply by ghost** [reply]
No problem at all, thank you!

**Reply by rsimon** [reply]
I did manage to publish a release after all :-)
https://www.npmjs.com/package/@annotorious/openseadragon-textlayer

**Reply by ghost** [reply]
That's great! I had a question about this plugin. In the example, it looks like the text layer is initialized on top of the viewer and text that is pre-written is displayed. A nice application indeed. But, does this plugin also support insertion of textboxes at the mouse position (when clicked) and entering text into the textbox from the keyboard?

**Reply by rsimon** [reply]
No, it only displays labels based on existing annotations.

**Reply by ghost** [reply]
Ok. Can I ask if this is present in annotorious v2 or if something similar to this is planned down the road?

**Reply by rsimon** [reply]
No this isn't present in Annotorious v2 either. (Annotorious 2 has the default popup for editing the annotation contents.)

Using the text layer directly to edit annotation content would be great, I agree. And arguably not so difficult perhaps? The text layer extension is just a bunch of absolutely positioned DIVs, so I guess it would be a matter of making them `contentEditable`, plus a bit of plumbing. 

I'd love to see it. But I'm not planning to build this myself at the moment. (Things can always change quickly if I happen to need it one of my own projects and/or find a paying customer. But it's not a requirement for any of my clients right now.)

**Reply by ghost** [reply]
Okay, thank you. I like the compact code in svelte for the ellipse. Would you say this functionality can be implemented in a similar way (I think storing the text is somewhat of an issue?)

**Reply by rsimon** [reply]
Yes, it would be a bit more involved I guess. Storing the text is the easy part. You'd just store it like any other annotation payload, in the annotation body. But the rest depends on how you want the interaction to work like. On click, you'd need to create a kind of stub annotation, and then use that as the container of your data. Annotorious 3 doesn't support point annotations yet, so you'd need to fake a rectangle annotation or something. Would take some experimenting to think it through.

**Reply by ghost** [reply]
Thanks!

**Reply by ghost** [reply]
Hi @rsimon. Merry Christmas! I had some questions about the internal workings of the library I thought you might be able to help me understand. 

I started by checking out the `@annotorious/annotorious` package alone, adding the EllipseEditor and RubberbandEllipse to an ellipse folder, modifying other scripts where ellipse is supposed to be present (alongside rectangle/polygon) and installing it separately. Wanted to see if `@annotorious/react` could detect the changes without the Annotorious plugin (which is working btw). But unfortunately, an error throws "No drawing tool ellipse".  I couldn't figure out why this is happening. Even the error "string" shows the same despite changing it. 

The reason for doing this is that I noticed that ellipse was already part of /annotation/shapes, /model/core and in some other places which I thought might also be needed for other shapes. Am I correct in assuming that any additional tool might also need to be added along with the other tools (in `@annotorious/annotorious`) apart from adding it to the Annotorious selector pack or are there other steps aside from this? 

Thanks.

**Reply by rsimon** [reply]
Thanks - and likewise! 🎄 

The tool plugin framework allows you to add new drawing tools __without__ having to fork the core library. Unless you are adding a new shape type, there would be no need to touch the `@annotorious/annotorious` package. (However, I assume that this is what you plan to do, so you'll probably need to look into that part later.)

Either way: new tools aren't automatically detected. You need to notify Annotorious about the presence of a new tool plugin. Here's where the Ellipse plugin registers itself with Annotorious:
https://github.com/annotorious/annotorious-v3-extensions/blob/main/annotorious-selector-pack/src/selectorPack.ts

If the tool isn't registered, it will cause the error message above. Also note that tools always come in two parts:

- First, the actual drawing tool, responsible for drawing a new shape
- Second, the "editor", responsible for modifying an existing shape.

In case of the rectangle, the tool provides that draggable rectangle shape, while the editor provides the box with handles for editing the shape once it's drawn.

Nonetheless: as you noticed, each shape primitive also needs to represented in the core somehow, so that common concerns (like computing the area of a polygon, doing mouse/shape intersection tests) don't have to be re-implemented over and over again by anyone building a new tool. That's the reason why there is also and "ellipse" folder in @annotorious/annotorious. But as you can see, it doesn't include anything related to the tools. Only a basic core data model and geometry helpers.

You would only need to make changes to the core if you want to, say, add a point or line drawing tool. The tools could still be implemented as plugins. But core geometry functions for point and line should be added to the core.

The thinking behind this is: core geometry types would be part of the framework. But different tools might allow people to draw shape primitives in different ways. (E.g. a freehand drawing tool for polygons, or a Photoshop-style magnetic lasso tool. Very different in terms of tool implementation. But still producing a polygon in the end.)

**Reply by ghost** [reply]
Thanks for this explanation! I sort of have a better idea about the workings but still stumbled upon some very inexplicable errors. I'll outline the steps I did below.

Checkout the `@annotorious/annotorious` package and **purposely** added ellipse portions to the core itself (meaning writing the EllipseEditor and RubberbandEllipse inside the core, and adding ellipse stuff along with rectangle, polygon where necessary), thereby not intending to use the selector-pack plugin. I think this should also mean that I have the option to choose "ellipse" along with "rectangle", "polygon" as part of default tools. 

I even changed some code (inside the src/Annotorious.ts) as shown in the image. Here I changed the string to display a slightly different error when it can't find the tool. This was done to see if this code actually runs when I change the drawingTool in the params.

![image](https://github.com/annotorious/annotorious/assets/23161642/2b0586c3-3b7c-46cb-a31e-68021cb8fa57)

Installed the package, and later installed `@annotorious/react`. 

The packages are installed just fine. The issue is when I change my tool to "ellipse", an error pops up saying, 
"No drawing tool named ellipse" even after changing the error string to "No such tool named {tool}" and also the console.log() code doesn't run. This makes me wonder if `@annotorious/react` even recognizes the changed `@annotorious/annotorious`? 

My intentions in going through with this is so that when I modify the core with Line or Point, the package allows me to.

**Reply by rsimon** [reply]
Just briefly:
- did you register "ellipse" in the list of available tools/editors?
- you need to `npm run build`. The React package picks up the __built__ core packages. Just changing the source code alone wouldn't affect the React package.

**Reply by ghost** [reply]
Yes, I put ellipse in the editors, tools directories. Also added it to the drawToolsRegistry.ts

![image](https://github.com/annotorious/annotorious/assets/23161642/1c2357d8-807e-4139-b86a-dc1bbda471e3)

And I added this `"prepare": "npm run build"` inside the scripts (in package.json). So when I `npm install`, it creates the dist folder. (Just a cross-check: I do see the ellipse editor in the dist).

**Reply by ghost** [reply]
If you'd like to have a look at the changes: https://github.com/ShashanKV98/annotorious-subdir/tree/dev

**Reply by rsimon** [reply]
Code looks OK the way you set it up. So it's likely a build problem. You could try running npm start in the annotorious package directly to confirm. (That would start the vanilla JS, non-OpenSeadragon version of Annotorious.) If you call listDrawingTools there, ellipse should come up in theory.

**Reply by ghost** [reply]
I ran as you suggested and ellipse shows up in the list. Also, I tried replacing the `<OpenSeadragonAnnotator tool ={tool}>` with `<ImageAnnotator>` and I'm able to draw an ellipse. I think I might know why I'm facing this issue. 

 `@annotorious/openseadragon` is in the peerDependenciesMeta of `@annotorious/react`'s package.json. In turn, the `@annotorious/openseadragon` has a dependency `"@annotorious/annotorious": "*",` which is probably installing the unmodified version. 
 
 I did notice that if I did not check out `@annotorious/react` and change its `"@annotorious/annotorious"` to my github URL, then even the `<ImageAnnotator>` wouldn't allow ellipse.

Relating to this, I noticed another thing. In the `@annotorious/openseadragon` package, inside the src/Annotorious.ts, there's this piece of code:
![image](https://github.com/annotorious/annotorious/assets/23161642/4f59bdf3-6c86-4f23-86dd-d16e136c7cb0)
This is the same error string which is thrown when choosing ellipse. (different from the slightly modified function in my checkedout  `@annotorious/annotorious`.

Later, I tried to checkout `@annotorious/openseadragon` too, in the hope that I could replace the `*` with my github URL and install it instead. But, I am stuck with many errors while runnning `npm run build`.

**Reply by ghost** [reply]
Sorry if this has gotten too long. Just want to add one last error message when running `npm run build` in the checked out `@annotorious/openseadragon`.
![image](https://github.com/annotorious/annotorious/assets/23161642/e1aa945b-1895-456d-a462-fc5b7579fd11)

P.S I removed a large number of typescript errors by adding "strict": false in tsconfig.json (felt wrong but couldn't do without it)

**Reply by rsimon** [reply]
I could never get rid of the last one myself. I  think it may be a Svelte issue. Depending on your version of TS, you may need to use // @ts-ignore on this one. FWIW its working for me with TS 4.9.5 without errors or warnings.

**Reply by ghost** [reply]
Ah, finally got this working! Thanks for the input.  I should've stayed with the versions already provided, since changing them to the latest got me all sorts of TS errors ( now I hate myself for doing it :\ ).

I did however notice that in the `@annotorious/openseadragon`, there's a file /src/annotation/pixi/stageRenderer.ts in which all the shapes were imported and used. If I had a new shape (like a Line), should I also be modifying this file? 

Also, came across the `presence` directory in the same package, with two scripts for Rectangle and Polygon. Since ellipse is working I presume that I won't need to change anything in there?

**Reply by rsimon** [reply]
Yes, exactly. The core needs to have knowledge about all shape primitives that tools could possibly create. That needs to be part of the core model, otherwise the renderers (WebGL or SVG) wouldn't know how to display them. That means eventually, things like Point, Line, Polyline etc. need to be added to the core model, and renderers (like the Pixi stageRenderer) need a bit of extra code to be able to render the shape.

Presence is a bit of an added thing: new shape types also need to be covered in the presence renderer. Ellipse is still missing, as you say. But I'll add it in due time. (FWIW the presence renderer isn't used by default. It's a utility for building GoogleDocs-like realtime apps, which will want an extra layer for indicating which other user in the room has selection on a given shape.)

**Reply by ghost** [reply]
Sounds like a great feature! Thanks again.

**Reply by ghost** [reply]
Hi @rsimon  I was trying to implement a freehand functionality (using `path` tag). It looked just fine for the non-openseadragon version (probably because it renders directly on the browser). But I am out of luck trying to get it to render on Pixi js for the OSD Viewer. 

I noticed that PIXI Graphics has APIs for rect, polygon and ellipse but for paths, functions like moveTo, quadraticCurveTo seem to be the closest. The issue with this is when there are sharp angles in the path ("V" shaped turns), PIXI fills the spaces with triangles thereby rendering a different shape. Have you ever come across such problems or happen to know what could be done to mitigate this? 
Thanks

**Reply by rsimon** [reply]
I haven't looked into how rendering of lines, polylines, curves etc. works in Pixi, TBH. (Beyond verifying that the do have `moveTo` and `lineTo` methods.) So can't really say much about it. Might this be something you can control via Pixi's fill- and line-style options somehow?

**Reply by ghost** [reply]
I tried changing the functions (beginFill, linestyle), even switching them to other places in the code, but wasn't helping much. 

I used a smoothing library to render the path. It outputs a path with a certain thickness although the "d' param in the path is a normal string of points. I did notice that if I didn't use the fill (or when made transparent), the additional polygons are not seen, but the path looks hollow (kind of like a pipe). I couldn't find any resource about rendering complex paths with Pixi. I guess people don't generally use it for a pen tool?

**Reply by rsimon** [reply]
Weird. WebGL is probably not the perfect environment for paths. But I'd be surprised if there was no way to do it at all. My hope was that `moveTo`/`lineTo` would allow drawing an un-closed path. And then it's up to the line style to set the thickness. 

In general, if you want to the thickness to be independent of the zoom factor, you can either use a path width of 1px (supported by the WebGL environment); or you can provide a px-width to to Pixi - but then have to handle scaling according to the current stage zoom factor yourself. (I'm doing this already for box/polygon outlines.)

Anyways: I'll try to find some time to experiment with this. So far, it's just me guessing what Pixi can and cannot do ;-)

**Reply by ghost** [reply]
Appreciate it, thanks! I tried with the same two functions of Pixi, but sometimes when the shape takes a few angles it weirdly closes the path itself. Attached is an image showing this behavior (third scribble from the top). 

![image](https://github.com/annotorious/annotorious/assets/23161642/6e57d542-dccc-4059-b491-16dc52c9627d)
